### PR TITLE
DOP-5132: Support image-less banners and add TypeScript support

### PR DESCRIPTION
**Pull request by rayangler**
### Stories/Links:

DOP-5132

### Current Behavior:

[Atlas prod](https://www.mongodb.com/docs/atlas/)
[Landing prod](https://www.mongodb.com/docs/)

### Staging Links:

[Atlas staging](https://docs-mongodb-org-stg.s3.us-east-2.amazonaws.com/sandbox/cloud-docs/raymund.rodriguez/DOP-5132-seamless-banner/index.html)
[Landing staging](https://docs-mongodb-org-stg.s3.us-east-2.amazonaws.com/master/landing/raymund.rodriguez/DOP-5132-seamless-banner/index.html)

### Notes:

* Adds support for customizing the banner text, pill text, and background color of the banner while keeping existing banner styles. Design has LGTM'd this [on Slack](https://mongodb.slack.com/archives/C04L4BVFEM9/p1733340866946199). For testing purposes, the `banner-dev` value on Atlas App Services has been updated with the new fields.
* This maintains backwards compatibility with image-based banners, in case there is a really custom design that needs to be hosted immediately.
* **ADDS INITIAL TYPESCRIPT SUPPORT.** I promised to start incrementally adding TS support whenever I touched components, so this is the start 😤. This does end up increasing the scope of the PR a bit, but I tried to limit the changes to initial setup (with linting, formatting, testing, etc.) and to the site banner components. This should **not** negatively impact current usage of the frontend. Just run `npm ci` and things should build as it does currently. **If it is preferred to split this into a separate PR, please let me know!**

### README updates

- - [ ] This PR introduces changes that should be reflected in the README, and I have made those updates.
- - [x] This PR does not introduce changes that should be reflected in the README

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -34,6 +40,8 @@
     "jest-environment-jsdom": "^29.5.0",
     "lint-staged": "^10.5.4",
     "prettier": "^2.2.1",
+    "ts-jest": "^29.2.5",
+    "typescript": "<5.4.0",
```
Comment (rayangler) [inline package.json:44]: Limiting the `typescript` version due to a bunch of incompatibility warnings and issues from `babel-eslint` and required TS linting parser/plugin. I tried to make config changes as minimal as possible for now, but we should have more flexibility on this once we get the chance to upgrade our linter (DOP-5057)

File: src/components/Banner/SiteBanner/index.tsx [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+  display: flex;
+  justify-content: center;
+  gap: 20px;
+  padding: 0 11px;
+  font-size: 13px;
+  line-height: 20px;
+
+  @media ${theme.screenSize.upToMedium} {
+    background-image: url(${getBannerSource(bannerContent.tabletImgPath)});
+    gap: 104px;
```
Comment (mmeigs) [inline src/components/Banner/SiteBanner/index.tsx:45]: Just curious on why tablet size has such a large gap

File: src/components/Banner/SiteBanner/index.tsx [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+const SiteBanner = () => {
+  const { bannerContent, setBannerContent } = useContext(HeaderContext);
+  const { snootyEnv } = useSiteMetadata();
+
+  useEffect(() => {
+    const fetchBannerContent = async () => {
+      try {
+        const res: SiteBannerContent = await fetchBanner(snootyEnv);
+        // Guard against missing banner content
+        if (res && (res.imgPath || res.text)) {
```
Comment (mmeigs) [inline src/components/Banner/SiteBanner/index.tsx:103]: nit: I'm wondering if we also want to guard against `res.url` not existing.

**Review by mmeigs — APPROVED** [review]
Ohhh this is very exciting about Typescript. Also, the banner looks great!

File: src/components/Banner/SiteBanner/index.tsx (hunk 3 above)
Comment (rayangler) [reply src/components/Banner/SiteBanner/index.tsx]: Good point! Added!

File: src/components/Banner/SiteBanner/index.tsx (hunk 2 above)
Comment (rayangler) [reply src/components/Banner/SiteBanner/index.tsx]: The Figma design showed a gap of 104px, but thinking about it now, seems like the intention is to just have the brand shape and pill be on the righthand side. I've updated the CSS to accommodate for max width on the text, while ensuring the flex container has `space-between` instead to make it more flexible. Thanks for bringing this up and let me know what you think!

File: src/types.ts
Comment (seungpark) [inline src/types.ts:1]: [nit] do we want to define this within SiteBanner/index or SiteBanner/types? seems a little loaded to put it into src/types

File: src/components/Banner/SiteBanner/index.tsx [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+  background-image: url(${getBannerSource(bannerContent.imgPath)});
+  background-position: center;
+  background-size: cover;
+  ${bannerContent.bgColor && `background-color: ${bannerContent.bgColor};`}
+  height: 100%;
+  display: flex;
+  justify-content: center;
+  gap: 20px;
+  padding: 0 11px;
+  font-size: 13px;
```
Comment (seungpark) [inline src/components/Banner/SiteBanner/index.tsx:40]: [minor nit] to use theme.fontSize here

File: src/components/Banner/SiteBanner/index.tsx [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+  @media ${theme.screenSize.upToMedium} {
+    background-image: url(${getBannerSource(bannerContent.tabletImgPath)});
+    justify-content: space-between;
+  }
+
+  @media ${theme.screenSize.upToSmall} {
+    background-image: url(${getBannerSource(bannerContent.mobileImgPath)});
+    gap: 28px;
+    font-size: 11px;
+    justify-content: space-between;
```
Comment (seungpark) [inline src/components/Banner/SiteBanner/index.tsx:52]: [minor nit] believe this is captured by upToMedium

File: src/components/Banner/SiteBanner/index.tsx [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+  }
+`;
+
+const bannerTextStyle = css`
+  align-self: center;
+  max-height: 40px;
+
+  @media ${theme.screenSize.upToMedium} {
+    max-width: 543px;
```
Comment (seungpark) [inline src/components/Banner/SiteBanner/index.tsx:61]: do we need these max widths? the flexbox + gap should handle the width of the text box

File: src/components/Banner/SiteBanner/index.tsx [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+  ${gridCell}
+  height: 40px;
+`;
+
+const pillStyle = css`
+  ${gridCell}
+  color: ${palette.green.dark3};
+  font-weight: 600;
+  line-height: 16px;
+  font-size: 12px;
```
Comment (seungpark) [inline src/components/Banner/SiteBanner/index.tsx:91]: nit for font size from theme

**Review by seungpark — COMMENTED** [review]
minor comments below but overall looks great! TS is very exciting! 🙌

File: src/components/Banner/SiteBanner/index.tsx (hunk 7 above)
Comment (rayangler) [reply src/components/Banner/SiteBanner/index.tsx]: Done

File: src/components/Banner/SiteBanner/index.tsx (hunk 6 above)
Comment (rayangler) [reply src/components/Banner/SiteBanner/index.tsx:59]: I think keeping the max width around would help make the max width guidelines included in the design more explicit vs. relying solely on flexbox/gap. Would it be preferable to remove?

File: src/components/Banner/SiteBanner/index.tsx (hunk 5 above)
Comment (rayangler) [reply src/components/Banner/SiteBanner/index.tsx]: Good point. Done

File: src/components/Banner/SiteBanner/index.tsx (hunk 4 above)
Comment (rayangler) [reply src/components/Banner/SiteBanner/index.tsx]: Done

File: src/types.ts
Comment (rayangler) [reply src/types.ts:1]: Makes sense. Done

File: src/components/Banner/SiteBanner/index.tsx (hunk 6 above)
Comment (seungpark) [reply src/components/Banner/SiteBanner/index.tsx:59]: minor preference, but don't feel too strongly about it!

Comment (rayangler) [reply src/components/Banner/SiteBanner/index.tsx]: Done

**Review by seungpark — APPROVED** [review]

**Pull request merged**